webapp/src/main/java/com/box/l10n/mojito/service/ai/translation/AITranslateJob.java

```diff
@@ -0,0 +1,257 @@
+
+  @Autowired LLMService llmService;
+
+  @Autowired MeterRegistry meterRegistry;
+
+  @Autowired RepositoryRepository repositoryRepository;
+
+  @Autowired RepositoryLocaleAIPromptRepository repositoryLocaleAIPromptRepository;
+
+  @Autowired AITranslationTextUnitFilterService aiTranslationTextUnitFilterService;
+
+  @Value("${l10n.ai.translation.pendingMT.expiryDuration:PT3H}")
+  Duration expiryDuration;
+
+  @Value("${l10n.ai.translation.reuseSourceOnLanguageMatch:false}")
+  boolean isReuseSourceOnLanguageMatch;
+
+  @Override
+  @Timed("AITranslateJob.call")
+  public Void call(AITranslateJobInput input) throws Exception {
+
+    try {
+      Repository repository =
+          repositoryRepository
+              .findById(input.getRepositoryId())
+              .orElseThrow(
+                  () ->
+                      new AITranslateJobException(
+                          "Repository with id " + input.getRepositoryId() + " not found."));
+
+      TmTextUnitPendingMT pendingMT =
+          tmTextUnitPendingMTRepository.findByTmTextUnitId(input.getTmTextUnitId());
+
+      if (pendingMT != null) {
+        if (!isExpired(pendingMT)) {
+          tmTextUnitRepository
+              .findById(input.getTmTextUnitId())
+              .ifPresent(
+                  tmTextUnit -> {
+                    if (aiTranslationTextUnitFilterService.isTranslatable(tmTextUnit, repository)) {
+                      translateLocales(input, tmTextUnit, repository);
+                    } else {
+                      logger.debug(
+                          "Text unit with name: {} should not be translated, skipping AI translation.",
+                          tmTextUnit.getName());
+                      meterRegistry.counter(
+                          "AITranslateJob.translate.notTranslatable",
+                          Tags.of("repository", repository.getName()));
+                    }
+                  });
+          meterRegistry
+              .timer("AITranslateJob.timeToMT", Tags.of("repository", repository.getName()))
+              .record(Duration.between(JSR310Migration.dateTimeNow(), pendingMT.getCreatedDate()));
+        } else {
+          // If the pending MT is expired, log an error and delete it
+          logger.error("Pending MT for tmTextUnitId: {} is expired", input.getTmTextUnitId());
+          meterRegistry.counter(
+              "AITranslateJob.expired", Tags.of("repository", repository.getName()));
+        }
+      } else {
+        logger.warn("No pending MT found for tmTextUnitId: {}", input.getTmTextUnitId());
+        meterRegistry.counter(
+            "AITranslateJob.noPendingMTFound", Tags.of("repository", repository.getName()));
+      }
+    } catch (Exception e) {
+      logger.error("Error running job for text unit id {}", input.getTmTextUnitId(), e);
+      meterRegistry.counter(
+          "AITranslateJob.error", Tags.of("repository", input.getRepositoryId().toString()));
+    } finally {
```

Comment:
To discuss: what happens if the translation call persistently fails on the pending TU, do we still mark it as "non-pending" here (delete it from the pending table)?

Is that the right flow for that TU?

Reply:
If retries get exhausted for a text unit translation then the AITranslateJob deletes the related `pendingMT` entity which frees that text unit up to be translated in the existing flow.

Reply:
As long as we can clearly track the number of terminal failures (i.e.: how many strings did not get LLM translated due to persistent errors) - that sounds good!